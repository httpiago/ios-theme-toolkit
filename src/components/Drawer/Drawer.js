import React from 'react'
import PropTypes from 'prop-types'
import { mapClassList } from '../../Utils'
import Portal from '../Portal/Portal'

/**
 * Drawer component.
 *
 * @since 0.5.0
 * @see https://httpiago.github.io/ios-theme-toolkit/#/components/Drawer
 */
export default class Drawer extends React.Component {
  drawerRef = React.createRef()

  componentDidUpdate(prevProps) {
    if (this.props.visible !== prevProps.visible) {
      if (this.props.visible === true) this.showDrawer()
      else this.hideDrawer()
    }
  }

  /**
   * Hide drawer with animation
   */
  hideDrawer = () => {
    this.drawerRef.current.style[this.styleProp] = ''
  }

  /**
   * Show drawer with animation
   */
  showDrawer = () => {
    const drawerElement = this.drawerRef.current
    const { placement } = this.props

    if (placement == 'right') {
      var styleProp = 'left'
      var value = document.body.offsetWidth - drawerElement.offsetWidth
    } else if (placement == 'left') {
      var styleProp = 'right'
      var value = document.body.offsetWidth - drawerElement.offsetWidth
    } else if (placement == 'top') {
      var styleProp = 'bottom'
      var value = document.body.offsetHeight - drawerElement.offsetHeight
    } else if (placement == 'bottom') {
      var styleProp = 'top'
      var value = document.body.offsetHeight - drawerElement.offsetHeight
    }

    // Save for be used in closeDrawer
    this.styleProp = styleProp

    drawerElement.style[styleProp] = value + 'px'
  }

  closeDrawer = () => {
    if (this.props.closeByBackdropClick === false) return

    this.props.onRequestClose()
  }

  render() {
    const { content, visible, placement, backdrop, className: aditionalClasses, onRequestClose, closeByBackdropClick, ...rest } = this.props

    const prefix = 'drawer'
    const classes = mapClassList({
      [`${prefix}--${placement}`]: true,
    })

    return (
      <Portal>
        {visible && backdrop && <div className="drawer__backdrop" onClick={this.closeDrawer} />}

        <div
          ref={this.drawerRef}
          className={`${prefix} ${classes} ${aditionalClasses} ${visible ? 'drawer--visible' : ''}`}
          {...rest}
        >
          {content}
        </div>
      </Portal>
    )
  }
}

Drawer.defaultProps = {
  className: '',
  placement: 'right',
  backdrop: true,
  closeByBackdropClick: true,
}

Drawer.propTypes = {
  /** Aditional classes. */
  className: PropTypes.string,
  /** The visibility state of drawer */
  visible: PropTypes.bool.isRequired,
  /** Drawer content */
  content: PropTypes.node.isRequired,
  /** Drawer placement */
  placement: PropTypes.oneOf(['top', 'left', 'bottom', 'right']),
  /** Function that will be called when the user clicks outside of modal and in the future when pressing the esc key */
  onRequestClose: PropTypes.func.isRequired,
  /** Show a transparent background behind the drawer */
  backdrop: PropTypes.bool,
  /** Close the drawer when user click in the backdrop */
  closeByBackdropClick: PropTypes.bool
}
