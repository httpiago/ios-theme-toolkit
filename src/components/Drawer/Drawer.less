@import '../../customizations.less';

// Default style
.drawer {
  background-color: #efeff4;
  padding: @base-padding (@base-padding + 5);
  box-sizing: border-box;
  position: fixed;
  display: flex;
  flex-direction: column;
  overflow: auto;
  width: initial;
  height: initial;
  font-family: @font-family;
  font-size: @font-size + 2;
  line-height: @base-line-height;
  z-index: 10015;
  -webkit-transition: left 218ms ease, top 218ms ease, bottom 218ms ease, right 218ms ease;
  transition: left 218ms ease, top 218ms ease, bottom 218ms ease, right 218ms ease;
}

.drawer__backdrop {
  background-color: rgba(0,0,0,0.6);
  position: fixed;
  height: 100%;
  width: 100%;
  top: 0px;
  left: 0px;
  z-index: 10000;
}

.drawer:not(.drawer--visible) {
  box-shadow: none !important;
}

.drawer--top {
  box-shadow: 0 2px 6px rgba(0,0,0,0.26);
  max-height: 80%;
  margin: 0 auto;
  right: 0;
  bottom: 100%;
  left: 0;
}

.drawer--left {
  box-shadow: 2px 0 6px rgba(0,0,0,0.26);
  max-width: 80%;
  margin: auto 0;
  width: 260px;
  bottom: 0;
  top: 0;
  right: 100%;
}

.drawer--bottom {
  box-shadow: 0 -2px 6px rgba(0,0,0,0.26);
  max-height: 80%;
  margin: 0 auto;
  right: 0;
  top: 100%;
  left: 0;
}

.drawer--right {
  box-shadow: -2px 0 6px rgba(0,0,0,0.26);
  max-width: 80%;
  margin: auto 0;
  width: 260px;
  left: 100%;
  top: 0;
  bottom: 0;
}

.drawer .list {
  margin-left: -(@base-padding + 5);
  margin-right: -(@base-padding + 5);
}
