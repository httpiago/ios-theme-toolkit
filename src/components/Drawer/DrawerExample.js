import React from 'react'
import Drawer from './Drawer'
import Button from '../Button/Button'
import List from '../List/List'

export default class DrawerExample extends React.Component {
  state = {
    drawer1Visibility: false,
    drawer2Visibility: false,
    drawer3Visibility: false,
    drawer4Visibility: false,
  }

  togglerDrawer = (id, val) => {
    let stateName = `drawer${id}Visibility`
    this.setState({ [stateName]: val })
  }

  render() {
    return (
      <React.Fragment>
        <Drawer
          visible={this.state.drawer1Visibility}
          onRequestClose={() => this.togglerDrawer(1, false)}
          content={(
            <React.Fragment>
              Veniam nulla esse tempor tempor et eu occaecat amet duis. Labore aliqua anim minim anim eiusmod tempor magna ea duis laborum minim do. Aliquip magna ullamco amet ipsum commodo mollit incididunt dolore incididunt magna. Ad nulla veniam elit ipsum veniam. Nisi enim ipsum exercitation ad dolor aliquip tempor id non incididunt aute officia.

              Pages:
              <List>
                <List.Item>Home page</List.Item>
                <List.Item>About</List.Item>
                <List.Item>Blog</List.Item>
                <List.Item>Contact</List.Item>
                <List.Item>Docs</List.Item>
                <List.Item>Configs</List.Item>
              </List>
            </React.Fragment>
          )}
        />

        <Button fill onClick={() => this.togglerDrawer(1, true)}>Open default drawer</Button>
        <br/><br/>

        <Drawer
          visible={this.state.drawer2Visibility}
          placement="left"
          onRequestClose={() => this.togglerDrawer(2, false)}
          content={(
            <React.Fragment>
              Veniam nulla esse tempor tempor et eu occaecat amet duis. Labore aliqua anim minim anim eiusmod tempor magna ea duis laborum minim do. Aliquip magna ullamco amet ipsum commodo mollit incididunt dolore incididunt magna. Ad nulla veniam elit ipsum veniam. Nisi enim ipsum exercitation ad dolor aliquip tempor id non incididunt aute officia.

              Pages:
              <List>
                <List.Item>Home page</List.Item>
                <List.Item>About</List.Item>
                <List.Item>Blog</List.Item>
                <List.Item>Contact</List.Item>
                <List.Item>Docs</List.Item>
                <List.Item>Configs</List.Item>
              </List>
            </React.Fragment>
          )}
        />

        <Button fill onClick={() => this.togglerDrawer(2, true)}>Open left drawer</Button>
        <br/><br/>

        <Drawer
          visible={this.state.drawer3Visibility}
          placement="bottom"
          onRequestClose={() => this.togglerDrawer(3, false)}
          content={'Veniam nulla esse tempor tempor et eu occaecat amet duis. Labore aliqua anim minim anim eiusmod tempor magna ea duis laborum minim do. Aliquip magna ullamco amet ipsum commodo mollit incididunt dolore incididunt magna. Ad nulla veniam elit ipsum veniam. Nisi enim ipsum exercitation ad dolor aliquip tempor id non incididunt aute officia.'}
        />

        <Button fill onClick={() => this.togglerDrawer(3, true)}>Open bottom drawer</Button>
        <br/><br/>

        <Drawer
          visible={this.state.drawer4Visibility}
          placement="top"
          onRequestClose={() => this.togglerDrawer(4, false)}
          content={'Veniam nulla esse tempor tempor et eu occaecat amet duis. Labore aliqua anim minim anim eiusmod tempor magna ea duis laborum minim do. Aliquip magna ullamco amet ipsum commodo mollit incididunt dolore incididunt magna. Ad nulla veniam elit ipsum veniam. Nisi enim ipsum exercitation ad dolor aliquip tempor id non incididunt aute officia.'}
        />

        <Button fill onClick={() => this.togglerDrawer(4, true)}>Open top drawer</Button>
        <br/><br/>
      </React.Fragment>
    )
  }
}
