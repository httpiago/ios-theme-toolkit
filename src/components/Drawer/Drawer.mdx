---
name: Drawer
route: /components/Drawer
menu: Componentes
---

import { Playground, PropsTable } from 'docz'
import { Drawer } from '../../index.js'
import '../../../dist/styles.css'
import DrawerExample from './DrawerExample'

# Drawer (Panel)

To use this component, simply import it into any file and pass the [modifiers](#/#props) in the properties.

Note: This component has no state (a controlled component), you must provide the visibility state of the component and close it "manually".

See more in the example below:

```jsx
import { Drawer, Button } from 'ios-theme-toolkit'

export default class DrawerExample extends React.Component {
  state = {
    drawerVisibility: false,
  }

  openDrawer = () => {
    this.setState({ drawerVisibility: true })
  }

  closeDrawer = () => {
    this.setState({ drawerVisibility: false })
  }

  render() {
    return (
      <React.Fragment>
        <Drawer
          visible={this.state.drawerVisibility}
          placement="right"
          style={{ width: 400 }}
          onRequestClose={this.closeDrawer}
        >
          Hello world
        </Drawer>

        <Button fill onClick={this.openDrawer}>Open drawer</Button>
      </React.Fragment>
    )
  }
}
```

## Examples

<Playground>
  <DrawerExample />
</Playground>

## Props

<PropsTable of={ Drawer } />

All the properties which are not listed above will be transferred to the \<div class="drawer"/ \> tag.
