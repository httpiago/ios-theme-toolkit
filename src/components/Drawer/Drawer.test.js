/**
 * TEST FILE
 * Please try writing the tests for all component behaviors,
 * manually call the functions within the class, change the properties...
 *
 * @see https://jestjs.io/docs/en/expect
 * @see https://github.com/airbnb/enzyme/tree/master/docs/api
 */
import React from 'react'
import { shallow } from 'enzyme'
import Drawer from './Drawer'

const COMPONENT = shallow(
  <Drawer
    visible={true}
    content={<div className="checkContent">Test content</div>}
    onRequestClose={() => {}}
    className="aditionalClass"
  />
)

describe('Test Drawer component', () => {
  it('should render correctly based on last snapshot', () => {
    expect(COMPONENT).toMatchSnapshot()
  })

})
