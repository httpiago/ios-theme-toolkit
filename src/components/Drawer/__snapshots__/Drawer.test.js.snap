// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Drawer component should render correctly based on last snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Drawer
    backdrop={true}
    className="aditionalClass"
    closeByBackdropClick={true}
    content={
      <div
        className="checkContent"
      >
        Test content
      </div>
    }
    onRequestClose={[Function]}
    placement="right"
    visible={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <div
          className="drawer__backdrop"
          onClick={[Function]}
        />,
        <div
          className="drawer drawer--right aditionalClass drawer--visible"
        >
          <div
            className="checkContent"
          >
            Test content
          </div>
        </div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "drawer__backdrop",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="checkContent"
          >
            Test content
          </div>,
          "className": "drawer drawer--right aditionalClass drawer--visible",
        },
        "ref": Object {
          "current": null,
        },
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Test content",
            "className": "checkContent",
          },
          "ref": null,
          "rendered": "Test content",
          "type": "div",
        },
        "type": "div",
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <div
            className="drawer__backdrop"
            onClick={[Function]}
          />,
          <div
            className="drawer drawer--right aditionalClass drawer--visible"
          >
            <div
              className="checkContent"
            >
              Test content
            </div>
          </div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "drawer__backdrop",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="checkContent"
            >
              Test content
            </div>,
            "className": "drawer drawer--right aditionalClass drawer--visible",
          },
          "ref": Object {
            "current": null,
          },
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Test content",
              "className": "checkContent",
            },
            "ref": null,
            "rendered": "Test content",
            "type": "div",
          },
          "type": "div",
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
